adr/0026-schema.md

```diff
@@ -0,0 +1,615 @@
+
+When Zarf introduces new keys that they are not ready to promise long term support for they will mark them as experimental in the schema. A key is assumed to be stable if it's not experimental or deprecated.
+
+There are several other keys we plan to deprecate with automated migrations to new fields
+- `.metadata.aggregateChecksum` -> `.build.aggregateChecksum`
+- Metadata fields `image`, `source`, `documentation`, `url`, `authors`, `vendors` -> will become a map of `annotations`
+- `noWait` -> `wait` which will default to true. This change will happen on both `.components.manifests` and `components.charts`
+- `yolo` -> `airgap` which will default to true
+- `.components.actions.[default/onAny].maxRetries` -> `.components.actions.[default/onAny].retries`
+- charts will change to avoid [current confusion with keys](https://github.com/defenseunicorns/zarf/issues/2245). Exactly one of the following field will exist for each `components.charts`.
+```yaml
+  helm:
+    url: https://stefanprodan.github.io/podinfo
+    name: podinfo # replaces repoName since it's only applicable for helm repos
+
+  git:
+    url: https://stefanprodan.github.io/podinfo
+    path: charts/podinfo
+
+  oci:
+    url: oci://ghcr.io/stefanprodan/charts/podinfo
+
+  local:
+   path: chart
+```
+- `actions.wait` will be deprecated in favor of a wait list per component that is run on deploy. `action.wait` blocks that are `onDeploy.After` can be auto migrated and will work if deployed on v1. `wait` blocks that are not on `onDeploy.After` will not be auto migrated and the package will error out if deployed on v1. Zarf v1 will use kstatus for wait functionality instead of `kubectl wait`. `apiVersion` will be added as a required field on `wait`, and the `apiVersion` will be automatically set to the most recent `apiVersion` if the waited for resource is not a custom resource. Users can still wait on custom resources, but the resources must implement kStatus. `wait.condition` will be removed from the schema and Zarf will assume it is waiting for `ready`. A `wait` block will not be auto migrated if `condition` is not `set` to  `ready`, `available` or equivalent.
+In summary, a `wait` block will be auto migrated if, and only if:
+  - it is `onDeploy.after`
+  - the condition is `ready`, `available` or equivalent
+  - The kind is not a custom resource.
+- There may be further, more holistic changes to how actions works within Zarf that would have a significant affect on the schema. This will be covered in a separate ADR.
+
+### BDD scenarios
+The following are (behavior driven development)[https://en.wikipedia.org/wiki/Behavior-driven_development] scenarios to provide context of what Zarf will do in specific situations given the above decisions.
```

Comment:
Since we do not (to my knowledge - happy to be wrong) practice BDD explicitly and/or learn on gherkin style testing I suggest we rewrite/replace the scenarios as a slightly distilled bit of peers for each of these scenarios.  I think what the scenarios describe is quite useful - I'm not certain BDD/gherkin language adds much and they *may* distract from the content.

Reply:
We encourage BDD in the [Zarf feature request template](https://github.com/defenseunicorns/zarf/issues/new?assignees=&labels=enhancement+%E2%9C%A8&projects=&template=feature_request.md&title=), but I recognize that not everyone is familiar with it. The substance of the scenarios is covered in the decisions tab, but I wanted to outline them more clearly as I believe they are what users will be most concerned with. Any suggestions on alternative formats?

Reply:
TIL!

For alternative formats I would say you could rewrite the first scenario as:

> For Zarf version v1 when the `zarf.yaml` has no apiVersion or deprecated keys when the user runs `zarf package create` they will receive an error and be told to run `zarf dev update-schema` or how to migrate off cosign key paths or how to use flavors over groups depending on the error.

Reply:
That said, my suggestion is not a *MUST* but more of a "maybe"